src/include/detail/graph/diskann.h

```diff
@@ -0,0 +1,144 @@
+  uint32_t npoints{0};
+  uint32_t ndim{0};
+
+  std::ifstream binary_file(path, std::ios::binary);
+  if (!binary_file.is_open()) {
+    throw std::runtime_error("Could not open file " + path);
+  }
+
+  binary_file.read((char*)&npoints, 4);
+  binary_file.read((char*)&ndim, 4);
+
+  auto x = ColMajorMatrix<float>(ndim, npoints);
+
+  binary_file.read((char*)x.data(), npoints * ndim * sizeof(float));
+  binary_file.close();
+
+  return x;
+}
+
+/**
+ * @brief Read a diskann index from disk
+ *
+ * @note: Currently this will read the test index that has a constant
+ * number of neighbors for each vertex.
+ * @todo: Implement loop to compare total bytes read to file size,
+ * cf DiskANN/src/in_mem_graph_store.cpp:135
+ */
+auto read_diskann_mem_index(const std::string& index) {
+  std::ifstream binary_file(index, std::ios::binary);
+  if (!binary_file.is_open()) {
+    throw std::runtime_error("Could not open file " + index);
+  }
+
+  uint64_t index_file_size;
+  uint32_t max_degree;
+  uint32_t medioid;
+  uint64_t vamana_frozen_num;
+
+  binary_file.read((char*)&index_file_size, 8);
+  binary_file.read((char*)&max_degree, 4);
+  binary_file.read((char*)&medioid, 4);
+  binary_file.read((char*)&vamana_frozen_num, 8);
+
+  size_t num_nodes = (index_file_size - 24) / (max_degree * 4 + 4);
+  auto g = detail::graph::index_adj_list<uint32_t>(num_nodes);
+
+  for (size_t node = 0; node < num_nodes; ++node) {
+    uint32_t num_neighbors;
+    binary_file.read((char*)&num_neighbors, 4);
+    for (size_t i = 0; i < num_neighbors; ++i) {
+      uint32_t id;
+      binary_file.read((char*)&id, 4);
+      g.add_edge(node, id);
+    }
+    binary_file.seekg(max_degree - num_neighbors, std::ios_base::cur);
+  }
+  binary_file.close();
+
+  return g;
+}
+
+auto read_diskann_mem_index_with_scores(
+    const std::string& index, const std::string& data) {
+  auto x = read_diskann_data(data);
+
+  // @todo get rid of copy pasta
+  std::ifstream binary_file(index, std::ios::binary);
+  if (!binary_file.is_open()) {
+    throw std::runtime_error("Could not open file " + index);
+  }
+
+  uint64_t index_file_size;
+  uint32_t max_degree;
+  uint32_t medioid;
+  uint64_t vamana_frozen_num;
+
+  binary_file.read((char*)&index_file_size, 8);
+  binary_file.read((char*)&max_degree, 4);
+  binary_file.read((char*)&medioid, 4);
+  binary_file.read((char*)&vamana_frozen_num, 8);
+
+  size_t num_nodes = (index_file_size - 24) / (max_degree * 4 + 4);
+  auto g = detail::graph::adj_list<float, uint32_t>(num_nodes);
+
+  for (size_t node = 0; node < num_nodes; ++node) {
+    uint32_t num_neighbors;
+    binary_file.read((char*)&num_neighbors, 4);
+    for (size_t i = 0; i < num_neighbors; ++i) {
+      uint32_t id;
+      binary_file.read((char*)&id, 4);
+      g.add_edge(node, id, sum_of_squares(x[node], x[id]));
+    }
+    // @todo ??? Is this right ???
+    binary_file.seekg(max_degree - num_neighbors, std::ios_base::cur);
```

Comment:
Looks like the changes you made on the other PR didn't make it over here: https://github.com/TileDB-Inc/TileDB-Vector-Search/pull/204/files#diff-66f0c6afcad02cfba5854fbdfd2241b0068368e87818d1b3cfc8e14428f99c62